nntrainer/app_context.cpp

```diff
@@ -136,18 +156,43 @@ void AppContext::setWorkingDirectory(const std::string &base) {
 }
 
 const std::string AppContext::getWorkingPath(const std::string &path) {
+  return getFullPath(path, working_path_base);
+}
 
-  /// if path is absolute, return path
-  if (path[0] == '/') {
-    return path;
-  }
+int AppContext::registerLayerPlugin(const std::string &path,
+                                    const std::string &base_path) {
+  const std::string full_path = getFullPath(path, base_path);
 
-  if (working_path_base == std::string()) {
-    return path == std::string() ? "." : path;
-  }
+  void *handle = dlopen(full_path.c_str(), RTLD_LAZY | RTLD_LOCAL);
+  const char *error_msg = dlerror();
+
+  NNTR_THROW_IF(handle == nullptr, std::invalid_argument)
+    << func_tag << "open plugin failed, reason: " << error_msg;
+
+  ml::train::LayerPluggable *pluggable =
+    reinterpret_cast<ml::train::LayerPluggable *>(
+      dlsym(handle, "ml_train_layer_pluggable"));
+
+  error_msg = dlerror();
+  auto close_dl = [handle] { dlclose(handle); };
+  NNTR_THROW_IF_CLEANUP(error_msg != nullptr, std::invalid_argument, close_dl)
+    << func_tag << "loading symbol failed, reason: " << error_msg;
+
+  auto layer = pluggable->createfunc();
+  auto type = layer->getType();
+  NNTR_THROW_IF_CLEANUP(type == "", std::invalid_argument, close_dl)
+    << func_tag << "custom layer must specify type name, but it is empty";
+  pluggable->destroyfunc(layer);
+
+  FactoryType<ml::train::Layer> factory_func =
+    [pluggable](const PropsType &prop) {
+      std::unique_ptr<ml::train::Layer> layer =
```

Comment:
I just wonder about the purpose of `PluggedLayer`, as `PluggedLayer` interface must be always same as ml::train::Layer (nntrainer::Layer for now).
How about just `std::unique_ptr<ml::train::Layer> layer = std::unique_ptr<nntrainer::Layer, decltype(&pluggable->destroyfunc)>(pluggable->createfunc(), &pluggable->destroyfunc)` or did I understood something wrong?

Reply:
That was my first try. The problem is `std::unique_ptr<nntrainer::Layer, decltype(&pluggable->destroyfunc)>` is different type from `std::unique_ptr<ml::train::Layer>`, so it does not play well with other unique_ptr. One option is we move every deleter to lambda but it is like x4 memory for a pointer, let alone change in a lot of spot.

Reply:
`std::unique_ptr<nntrainer::Layer, decltype(&pluggable->destroyfunc)>` should be able to reinterpret as `std::unique_ptr<ml::train::Layer>`??

Reply:
> std::unique_ptr<nntrainer::Layer, decltype(&pluggable->destroyfunc)> should be able to reinterpret as std::unique_ptr<ml::train::Layer>??

This won't call `destroyfunc()` as it is reinterpreted. thus not viable choice

> Additionally, std::unique_ptr<nntrainer::Layer, decltype(&pluggable->destroyfunc)> will anyway be replaced with std::unique_ptr<ml::train::Layer, decltype(&pluggable->destroyfunc)> once ml::train::Layer interface is finalized.

It doesn't have to be in this setup in the cost of pimpl delegation. Downside of changing `std::unique_ptr<Layer>` to `std::unique_ptr<Layer, lambda>`, This affects the performance let alone usability (it affects usability because you always need `std::unique_ptr<Layer, std::function...>` in every 
signature and variables.)

you might wanna read https://dev.krzaq.cc/post/you-dont-need-a-stateful-deleter-in-your-unique_ptr-usually/

Reply:
If it can't, IMO, it would be better to have an alias for `std::unique_ptr<ml::train::layer, lambda>` and use it, than maintaining another Layer extension `PluggedLayer` and then `PluggedOptimizer`, `PluggedModel` and so on.

Reply:
> If it can't, IMO, it would be better to have an alias for std::unique_ptr<ml::train::layer, lambda> and use it, than maintaining another Layer extension PluggedLayer and then PluggedOptimizer, PluggedModel and so on.

plus, the unique_ptr won't be default constructible 
```cpp
using PtrType<Layer> = std::unique_ptr<ml::train::layer, lambda>
std::vector<PtrType<Layer>> // this will be compile error as PtrType<Layer> is not default constructible.
```